processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -0,0 +1,195 @@
+// limitations under the License.
+
+package metricstransformprocessor
+
+import (
+	"context"
+	"log"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+)
+
+const (
+	validNewNameFuncName  = "validNewName()"
+	validNewLabelFuncName = "validNewLabel()"
+)
+
+type metricsTransformProcessor struct {
+	cfg        *Config
+	next       consumer.MetricsConsumer
+	transforms []Transform
+}
+
+var _ component.MetricsProcessor = (*metricsTransformProcessor)(nil)
+
+func newMetricsTransformProcessor(next consumer.MetricsConsumer, cfg *Config) (*metricsTransformProcessor, error) {
+	return &metricsTransformProcessor{
+		cfg:        cfg,
+		next:       next,
+		transforms: cfg.Transforms,
+	}, nil
+}
+
+// GetCapabilities returns the Capabilities associated with the metrics transform processor.
+func (mtp *metricsTransformProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (*metricsTransformProcessor) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (*metricsTransformProcessor) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+// ConsumeMetrics implements the MetricsProcessor interface.
+func (mtp *metricsTransformProcessor) ConsumeMetrics(ctx context.Context, md pdata.Metrics) error {
+	return mtp.next.ConsumeMetrics(ctx, mtp.transform(md))
+}
+
+// transform transforms the metrics based on the information specified in the config.
+func (mtp *metricsTransformProcessor) transform(md pdata.Metrics) pdata.Metrics {
+	mds := pdatautil.MetricsToMetricsData(md)
+
+	for i, data := range mds {
+		nameToMetricMapping := make(map[string]*metricspb.Metric)
+		// O(len(data.Metrics))
+		for _, metric := range data.Metrics {
+			nameToMetricMapping[metric.MetricDescriptor.Name] = metric
+		}
+
+		for _, transform := range mtp.transforms {
+			if !mtp.validNewName(transform, nameToMetricMapping) {
+				log.Printf("error running %q processor due to collision %q: %v with existing metric names detected by the function %q", typeStr, NewNameFieldName, transform.NewName, validNewNameFuncName)
+				continue
+			}
+
+			metric, ok := nameToMetricMapping[transform.MetricName]
+			if !ok {
+				continue
+			}
+
+			// mtp.action is already validated to only contain either update or insert
+			if transform.Action == Update {
+				mtp.update(metric, transform)
+				if transform.NewName == "" {
+					continue
+				}
+				// if name is updated, the map has to be updated
+				nameToMetricMapping[transform.NewName] = nameToMetricMapping[transform.MetricName]
+				delete(nameToMetricMapping, transform.MetricName)
```

Comment:
I thought the order of transformations here should matter because when I discussed with Dave and Quentin, the transforms list should be an ordered action list that will perform the transforms one by one to the metrics. Do you think that this shouldn't be the case?